File: Skydown App/Services/Payments/MembershipAnalyticsTracker.swift
import Foundation
#if canImport(FirebaseAnalytics)
import FirebaseAnalytics
#endif
#if canImport(FirebaseFunctions)
import FirebaseFunctions
#endif

struct MembershipAnalyticsTracker {
    func track(
        _ event: String,
        platform: String = "ios",
        reason: String? = nil,
        plan: String? = nil,
        annual: Bool? = nil,
        surface: String? = nil,
        currentPlan: String? = nil
    ) {
#if canImport(FirebaseAnalytics)
        var payload: [String: Any] = ["platform": platform]
        if let reason { payload["reason"] = reason }
        if let plan { payload["plan"] = plan }
        if let annual { payload["annual"] = annual ? "true" : "false" }
        if let surface { payload["surface"] = surface }
        if let currentPlan { payload["currentPlan"] = currentPlan }
        Analytics.logEvent(event, parameters: payload)
#endif
#if canImport(FirebaseFunctions)
        var mirrorPayload: [String: Any] = [
            "eventName": event,
            "platform": platform
        ]
        if let reason { mirrorPayload["reason"] = reason }
        if let plan { mirrorPayload["plan"] = plan }
        if let annual { mirrorPayload["annual"] = annual }
        if let surface { mirrorPayload["surface"] = surface }
        if let currentPlan { mirrorPayload["currentPlan"] = currentPlan }
        Task {
            do {
                _ = try await Functions.functions(region: "us-central1")
                    .httpsCallable("recordAiMembershipEvent")
                    .call(mirrorPayload)
            } catch {
                // Keep analytics fire-and-forget; failures should not impact caller flow.
            }
        }
#endif
    }
}

File: androidApp/src/main/java/com/nash/skyos/data/MembershipAnalyticsTracker.kt
package com.nash.skyos.data

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.functions.FirebaseFunctions

class MembershipAnalyticsTracker(
    context: Context,
) {
    private val analytics = FirebaseAnalytics.getInstance(context)
    private val functions = FirebaseFunctions.getInstance("us-central1")

    fun track(
        event: String,
        platform: String = "android",
        reason: String? = null,
        plan: String? = null,
        annual: Boolean? = null,
        surface: String? = null,
        currentPlan: String? = null,
    ) {
        val params = Bundle().apply {
            putString("platform", platform)
            reason?.let { putString("reason", it) }
            plan?.let { putString("plan", it) }
            annual?.let { putString("annual", it.toString()) }
            surface?.let { putString("surface", it) }
            currentPlan?.let { putString("currentPlan", it) }
        }
        analytics.logEvent(event, params)
        val payload = hashMapOf<String, Any>(
            "eventName" to event,
            "platform" to platform,
        )
        reason?.let { payload["reason"] = it }
        plan?.let { payload["plan"] = it }
        annual?.let { payload["annual"] = it }
        surface?.let { payload["surface"] = it }
        currentPlan?.let { payload["currentPlan"] = it }
        functions
            .getHttpsCallable("recordAiMembershipEvent")
            .call(payload)
    }
}
